### PR TITLE
Refactor replace RzBinElfDynamicInfo by RzBinElfDtDynamic ##refactor

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

The old struct (RzBinElfDynamicInfo) had a bad interface.

* removed RzBinElfDynamicInfo
* centralized all dt_dynamic information
* use opaque struct for RzBinElfDtDynamic
* use hash table to store informations

**There is a warning about eprintf, I don't know how to remove it.**

```
../librz/include/rz_types.h:402: warning: "eprintf" redefined
  402 | #define eprintf(...) fprintf(stderr, __VA_ARGS__)
      |
In file included from ../subprojects/sdb/src/ls.h:5,
                 from ../subprojects/sdb/src/ht_inc.h:49,
                 from ../subprojects/sdb/src/ht_uu.h:9,
                 from ../librz/bin/format/elf/elf_dynamic.h:5,
                 from ../librz/bin/format/elf/elf_dynamic.c:5:
../subprojects/sdb/src/types.h:13: note: this is the location of the previous definition
   13 | #define eprintf(...) fprintf(stderr,__VA_ARGS__)
      |
[618/1610] Compiling C object librz/bin/librz_bin.so.0.3.0-git.p/format_elf_elf64_dynamic.c.o
In file included from ../librz/bin/format/elf/elf.h:6,
                 from ../librz/bin/format/elf/elf_dynamic.h:6,
                 from ../librz/bin/format/elf/elf_dynamic.c:5,
                 from ../librz/bin/format/elf/elf64_dynamic.c:6:
../librz/include/rz_types.h:402: warning: "eprintf" redefined
  402 | #define eprintf(...) fprintf(stderr, __VA_ARGS__)
      |
In file included from ../subprojects/sdb/src/ls.h:5,
                 from ../subprojects/sdb/src/ht_inc.h:49,
                 from ../subprojects/sdb/src/ht_uu.h:9,
                 from ../librz/bin/format/elf/elf_dynamic.h:5,
                 from ../librz/bin/format/elf/elf_dynamic.c:5,
                 from ../librz/bin/format/elf/elf64_dynamic.c:6:
../subprojects/sdb/src/types.h:13: note: this is the location of the previous definition
   13 | #define eprintf(...) fprintf(stderr,__VA_ARGS__)
      |
```

**Test plan**
```
rz-test
```

**Closing issues**

...

### CODE REVIEW

File: librz/bin/format/elf/elf_dynamic.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#include "elf.h"
+
+#ifndef _INCLUDE_ELF_DYNAMIC_H_
+#define _INCLUDE_ELF_DYNAMIC_H_
+
+typedef struct Elf_(rz_bin_elf_dt_dynamic) {
+	HtUU *info;
+	RzVector *dt_needed;
+}
+RzBinElfDtDynamic;
```
Comment (ret2libc) [inline librz/bin/format/elf/elf_dynamic.h]:
```suggestion
} RzBinElfDtDynamic;
```

**Review by ret2libc — COMMENTED** [review]
What's the advantage of having an opaque structure here?

**Comment by ghost** [comment]
> What's the advantage of having an opaque structure here?

The user shouldn't use struct members:
* Even if the implementation is modified the API doesn't change.
* By design the dev will always use the API
* ht_uu_find API is bad, and the user shouldn't use it
* Elf_(rz_bin_elf_has_dt_info) should always be preferred to bin->dt_dynamic && bin->dt_dynamic->info && bin->dt_dynamic->dt_needed

**Comment by ret2libc** [comment]
> > What's the advantage of having an opaque structure here?
> 
> The user shouldn't use struct members:
> 
>     * Even if the implementation is modified the API doesn't change.
> 
>     * By design the dev will always use the API

I think in this case these benefits are very small, given that the elf bin plugin is not really an exposed API anyway, but just an internal detail. So using fields or a method to access everything does not really change too much.

> 
>     * ht_uu_find API is bad, and the user shouldn't use it

How is that bad? Asking genuinely since it is used everywhere.

**Comment by ghost** [comment]
> `* Even if the implementation is modified the API doesn't change.`



> > > What's the advantage of having an opaque structure here?
> > 
> > 
> > The user shouldn't use struct members:
> > ```
> > * Even if the implementation is modified the API doesn't change.
> > 
> > * By design the dev will always use the API
> > ```
> 
> I think in this case these benefits are very small, given that the elf bin plugin is not really an exposed API anyway, but just an internal detail. So using fields or a method to access everything does not really change too much.

The main problem of librz/bin/format/elf/* is the lack of harmonization. Forcing a dev to use the "API" will make the overall code base cleaner. It is maybe a small immediate benefits, but the overall quality of the code base will go up over time (with more and more opaque struct and such "API").

> > ```
> > * ht_uu_find API is bad, and the user shouldn't use it
> > ```
> 
> How is that bad? Asking genuinely since it is used everywhere.

The interface is `ut64 ht_uu_find(HtUU *ptr, ut64 key, bool *found)` which make no sense. When the function query a lot of variable, it is messy.

```
bool found;

ut64 addr = ht_uu_find(bin->dt_dynamic, DT_PLTGOT, &found);

if (!found) {
	// fail
}

// success
```

vs

```
ut64 addr;

if (!Elf_(rz_bin_elf_get_dt_info)(bin, DT_PLTGOT, &addr)) {
	// fail
}

// success
```

File: librz/bin/format/elf/elf_dynamic.h (hunk 1 above)
Comment (ghost) [reply librz/bin/format/elf/elf_dynamic.h]: It was clang format. But I remove the Elf_() macro (unused) and format again.

**Comment by ret2libc** [comment]
> > > ```
> > > * ht_uu_find API is bad, and the user shouldn't use it
> > > ```
> > 
> > 
> > How is that bad? Asking genuinely since it is used everywhere.
> 
> The interface is `ut64 ht_uu_find(HtUU *ptr, ut64 key, bool *found)` which make no sense. When the function query a lot of variable, it is messy.
> 
> ```
> bool found;
> 
> ut64 addr = ht_uu_find(bin->dt_dynamic, DT_PLTGOT, &found);
> 
> if (!found) {
> 	// fail
> }
> 
> // success
> ```
> 
> vs
> 
> ```
> ut64 addr;
> 
> if (!Elf_(rz_bin_elf_get_dt_info)(bin, DT_PLTGOT, &addr)) {
> 	// fail
> }
> 
> // success
> ```

Well, let's say that in this particular case that looks indeed better :D but for a generic API I think Ht is ok.

**Comment by ghost** [comment]
Do you have any idea how to fix the warning?

File: librz/bin/format/elf/elf_dynamic.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// SPDX-FileCopyrightText: 2021 08A <08A@riseup.net>
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-License-Identifier: LPGL-3.0-only
+
+#include <ht_uu.h>
+#include "elf.h"
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/elf_dynamic.h:5]: Swapping these fixes the `eprintf()` warning (the sdb one contains an explicit `#undef` so it works if that comes later)

File: librz/bin/format/elf/elf_dynamic.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+#ifndef _INCLUDE_ELF_DYNAMIC_H_
+#define _INCLUDE_ELF_DYNAMIC_H_
+
+typedef struct rz_bin_elf_dt_dynamic {
+	HtUU *info;
+	RzVector *dt_needed;
+} RzBinElfDtDynamic;
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/elf_dynamic.h]:
```
../librz/bin/format/elf/elf_dynamic.h:14:3: warning: redefinition of typedef 'RzBinElfDtDynamic' is a C11 feature [-Wtypedef-redefinition]
} RzBinElfDtDynamic;
  ^
../librz/bin/format/elf/elf.h:159:38: note: previous definition is here
typedef struct rz_bin_elf_dt_dynamic RzBinElfDtDynamic; // elf_dynamic.h
                                     ^
```

You should be able to fix this by removing the typedef here and just defining the struct alone:
```c
struct rz_bin_elf_dt_dynamic {
    HtUU *info;
    RzVector *dt_needed;
};
```

**Review by thestr4ng3r — COMMENTED** [review]
overall looks good to me.

**Review by thestr4ng3r — APPROVED** [review]

**Comment by ghost** [comment]
I forgot to add @XVilka

**Review by XVilka — APPROVED** [review]
Code looks good. What is scary though lines like:
```c
 static size_t populate_relocs_record_from_dynamic_aux(ELFOBJ *bin, RzBinElfReloc *relocs, size_t pos, size_t num_relocs, ut64 addr, ut64 size, ut64 entry_size, ut64 rel_mode) {
 ```
 
 I suggest to avoid ultra-long lines or too much function parameters when possible.

**Comment by ghost** [comment]
> Code looks good. What is scary though lines like:
> 
> ```c
>  static size_t populate_relocs_record_from_dynamic_aux(ELFOBJ *bin, RzBinElfReloc *relocs, size_t pos, size_t num_relocs, ut64 addr, ut64 size, ut64 entry_size, ut64 rel_mode) {
> ```
> 
> I suggest to avoid ultra-long lines or too much function parameters when possible.

Yes, I am not proud of this.

**Pull request merged**